[v0.1] Basic Count Statistics - Item Counter implemented

lucasferrari07 opened the pull request:
Function itemCounts returns a Map with the count of each element in the array received as parameter. For every element in the array, if it already exists in the Map, the code store it's current value to be incrementedin the Map. Otherwise, it is stored with a initial value of 1

lucasferrari07 committed:
[v0.1] Basic Count Statistics - Item Counter implemented

Function itemCounts returns a Map with the count of each element in the array received as parameter. For every element in the array, if it already exists in the Map, the code store it's current value to be incrementedin the Map. Otherwise, it is stored with a initial value of 1

lucasferrari07 commented on textalyze.js:
```diff
@@ -14,8 +14,10 @@
 
-  // Your code here.
-  // Run 'npm test' to see what tests need to pass.
+  for (let a of array) {
```
I now see that `a` should be a const as it's value will not be changed.

lucasferrari07 committed:
[v0.1] Basic Count Statistics - printing the elements to the console

Implemented printing, so we can see in the console the elements and their counts

lucasferrari07 committed:
Fixed const usage on for statement and removed logging from itemCounts function

jfarmer commented on textalyze.js:
```diff
@@ -14,10 +6,17 @@
 
-  // Your code here.
-  // Run 'npm test' to see what tests need to pass.
+  for (const a of array) {
```
Exercise: how would you write this as a single call to [reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce)?

jfarmer commented on textalyze.js:
```diff
@@ -14,10 +6,17 @@
   return counts;
 }
 
+const elements = ["a", "a", "a", "b", "b", "c"];
```
We'd like this file to be both included as a module and run via `npm start`.  Check out https://stackoverflow.com/questions/4981891/node-js-equivalent-of-pythons-if-name-main for how to do this.

You want

```javascript
if (require.main == module) {
  const array = ['a', 'a', 'a', ...];
}
```

jfarmer replied:
Make sure you're using ESlint.  I've set up the project to automatically enforce a style guide using AirBnb's recommended style.

You can set up your editor to do it automatically or you can manually run `npx eslint` to run the version included with the project on the command line.

jfarmer commented on textalyze.js:
```diff
@@ -14,10 +6,17 @@
 
+const elements = ["a", "a", "a", "b", "b", "c"];
+console.log(`Count for [${elements}] is...`);
+itemCounts(elements).forEach((count, elem) => {
```
Try to be consistent — you use `forEach` here but `for ... in ...` above.  The AirBnb style guide will tell you to use `forEach`.

jfarmer commented on textalyze.js:
```diff
@@ -14,10 +6,17 @@
+console.log(`Count for [${elements}] is...`);
+itemCounts(elements).forEach((count, elem) => {
+  console.log(elem, count);
+});
```
Check out [console.table](https://developer.mozilla.org/en-US/docs/Web/API/Console/table) — it's useful for quickly printing out tables in debugging situations.

jfarmer reviewed (approved):
Try to write good commit messages and good PR messages. A good commit message means that someone else could read the commit message and understand what is happening.

Read:
- https://chris.beams.io/posts/git-commit/
- https://github.com/erlang/otp/wiki/writing-good-commit-messages
- https://medium.com/@andrewhowdencom/anatomy-of-a-good-commit-message-acd9c4490437

Rather than saying "pull request for first iteration" or "implemented itemCounts" without anything else, treat it like an actual project.  Commit messages and PR messages are a kind of documentation.

You want to say something like:

```
Implement itemCounts

itemCounts takes in an Array and returns a Map containing the count
of each item in the Array.  For example, itemCounts(['a', 'a', 'b']) would
return the Map { 'a': 2, 'b': 1 }.
```

The sooner you get into this habit the sooner it will stop feeling like homework.

Later, when the projects become more substantial, we will _need_ good commit messages and PR descriptions in order to give you a quick, thorough code review.

lucasferrari07 committed:
Implement suggested changes for PR #1

- Changed the array.forEach approach for the array.reduce one, as it's now a single call.
- Isolated the execution part of the file so it runs only if the module is the main module.

lucasferrari07 committed:
Implement suggested changes for PR #1

- Changed the  approach for the  one, as it's now a single call.
- Isolated the execution part of the file so it runs only if the module is the main module.

lucasferrari07 committed:
Merge branch 'basic-count-statistics' of github.com:lucasferrari07/text-analysis-js into basic-count-statistics
